### DISCUSSION TITLE
2023 Roadmap

### DISCUSSION THREAD

**Discussion by ryanmoran** [paketo-buildpacks/feedback#58] [Ideas]
With 2022 fast approaching its end, we want to take the time to gather ideas for the next year's roadmap. 2022 was a super productive year for the project and we got a lot done.

What do you think the Paketo Buildpacks project should be working on in 2023? Check out last years [discussion](https://github.com/orgs/paketo-buildpacks/discussions/24) and [roadmap](https://blog.paketo.io/posts/2022-roadmap/) for inspiration. Let us know by commenting and upvoting items below.

After the new year, the Steering Committee will synthesize this discussion into a set of themes for the roadmap in 2023. We look forward to hearing what great new things Paketo Buildpacks should be doing next year!

Thank you!

**Comment by loewenstein** [comment]
I would like to see the Node.js language family to graduate to v1.0.

**Reply by robdimsdale** [reply]
For this specific topic, you can follow along with the tracking issue here: https://github.com/paketo-buildpacks/nodejs/issues/680

**Comment by loewenstein** [comment]
Currently looking into the [2022 roadmap discussion](https://github.com/orgs/paketo-buildpacks/discussions/24#discussioncomment-2312032) as a source for inspiration: Has [Java RFC 0007](https://github.com/paketo-buildpacks/rfcs/blob/main/text/java/0007-low-profile-memory-calc.md) been implemented in the meantime? If not, should it be?

**Reply by dmikusa** [reply]
The RFC was approved, but not implemented. I am still in favor of it, but I think it's slipped in priority because it will be a decent amount of work and we haven't had a lot of requests for it.

tl;dr - this RFC is to adjust the memory calculator to work better when less than 1G of RAM is given. If folks are interested in this being prioritized, please up-vote this comment.

**Comment by loewenstein** [comment]
Separating buildpacks and dependency management. See [Dan Mikusa’s work in progress proposal](https://docs.google.com/document/d/1g5rRW-oE_v8Gdvz-CiCOK9z2rxg6L5XniKI25Zq2j6M/edit#).

**Note**: We've decided in 2022 that we won't push for it, as there are other means to reach our goals. But if there would be pull from the community, we would likely be able to help.

**Comment by loewenstein** [comment]
Has the [Remove Stacks & Mixins RFC](https://github.com/buildpacks/rfcs/blob/main/text/0096-remove-stacks-mixins.md) been tackled or at least planned yet?

**Note**: If we go for it, I would like to additionally see less use of the `tiny` stack ID. The most prominent example I see that is imho worth removing is the check for `tiny` to decide if a `bash` is available. Maybe together with the [Build config RFC](https://github.com/buildpacks/rfcs/blob/main/text/0109-build-config.md) this could be delegated to be a user choice that gets defaulted in the builder?

**Comment by loewenstein** [comment]
Should we merge `packit` and `libpak` to avoid having to maintain two Go packages?

**Reply by dmikusa** [reply]
We might want to consider trying to push for a libcnb v2 release this coming year. It is somewhat out of our control as it's upstream, but if we ever want to get to a single library that is the first step. The v2 libcnb library is being modified so that it can be the base for both packit and libpack, which means the core buildpack spec implementation will be the same for both. What's left after that in packit and libpak will be conventions/opinions held by those libraries.

**Comment by dmikusa** [comment]
I would like to see the next step in regard to ARM64 images. In 2022, we took some steps, making it possible for users to build their own ARM64 images and that resulted in an unofficial ARM64 image for Java users. I'd like to see more in 2023.

I'd like to see at least one Paketo ARM64 stack available (build and run images). The base would probably be a good place to start as it supports a decent number of buildpacks. If we could get tiny & full too, that would be awesome. That would take some of the work away from creating your own ARM64 images. I don't think it would require upstream tool changes either, so it seems feasible.

**Reply by joshlong** [reply]
Please, please support ARM. I don't trust software I can't preflight on my own machines and ARM has been around in some incarnation or another for years, most recently with apple's m1.

**Reply by DeanPoulin** [reply]
I went to the AWS re:Invent conference this year and they're really pushing their new Arm64 based Graviton3 processors for the benefit of higher performance and lower cost.

We're planning to move our Spring Boot based AWS workloads over to ARM in 2023. Having paketo-buildpacks capable of producing ARM images would be tremendously helpful with that migration.

**Comment by loewenstein** [comment]
Not sure, but would it make sense to introduce stacks for 22.10, 23.04 and 23.10 as a preparation for quick adoption of 24.04?

Like we could treat it similarly to e.g. non-LTS versions of Java and Node.js.

**Comment by sophiewigmore** [comment]
I believe auto-generated reference docs are already on our 2022 Roadmap / the core team has discussed doing this recently. I wanted to capture my desire for this here as well.

**Comment by sophiewigmore** [comment]
I'd like to finally align on a path forward for debug logging across the project in both Java/non-Java buildpacks.

**Reply by dmikusa** [reply]
Question for clarification. Are you looking to see [RFC 0027](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0027-log-levels.md) implemented? Or is this more than what's covered in that RFC?

**Reply by sophiewigmore** [reply]
A step further than what's in that RFC, actually aligning on a standard for what should be included in debug logs. Right now, some of the non-Java buildpacks have debug logs, but they're very inconsistent in the content they provide. I'd love to have a more uniform experience

**Comment by dmikusa** [comment]
Solve the layer limit problem in the Paketo Builders. Probably by squashing layers in some way like per language family or composite buildpack. However, we could also consider breaking away from the all-in-one builder and exploring language-specific builders instead (do feel free to comment here with thoughts on this).

This is presently stifling things the projects would like to do, such as adding the APM-related buildpacks to the builders or adding more JVM providers to the default Java composite buildpack.

**Reply by dashaun** [reply]
What is the limit, where is it being hit?  I like the idea of language-specific builders.  In all of the places that I use buildpacks, I have the option to add multiple, and change the order, etc.  What are some of the reasons to continue supporting the all-in-one builder approach?

**Reply by kevin-ortega** [reply]
I really like the idea of language-specific builders.

**Reply by dmikusa** [reply]
There's a hard limit of 127 layers. It is a Linux kernel limit, so it cannot be changed.

The rationale for an all-in-one builder are a.) convenience, you don't have to think about it, it's always the same one, b.) discoverability, `pack builder suggest` lists some builders, I'm not sure if we could put every language in there, c.) it's easier, especially for new users, everyone uses the same builder and it just works.

The language-specific builders, IMO, would be more of a power-user feature. If you're familiar with buildpacks and know what buildpacks you need, a language-specific builder would be more streamlined. Fewer layers, it would download faster, and take up less space, but still give you the functionality you need.

**Reply by till** [reply]
Agree on the builders. While I also like language specific builders, it sounds like an optimization for a later stage.

For first user experience, if we can shrink the full builder to make that experience a bit more snappy, that would be something. Not sure if there's low hanging fruit.

**Comment by sambhav** [comment]
I would be interested in providing buildpacks to create dev images. i.e. images that can be used as a vscode dev container for eg.

**Comment by sambhav** [comment]
Move dependencies to upstream binaries wherever possible. Currently the Java buildpacks try to directly include upstream dependencies wherever possible. The other buildpacks however seen to rely on the dependencies served by the paketo dep server. In some cases, this is required (for eg. python and ruby interpreters) but there are cases where this is not necessary. In general I would like to see the entire paketo buildpack distribution/dep-server potentially updated so that the dependencies we custom build are more easily discoverable from the relevant buildpack and we upstream sources.

**Reply by sophiewigmore** [reply]
@samj1912 the good news on this one is that we're almost done moving away from the dep-server already, moving a model that A. tries to use upstream dependencies where possible, and B. has discoverable code alongside the buildpack itself for version discovery, metadata generation, and compilation (in cases where we have to do it). It was a huge undertaking, but we have an issue here for this: https://github.com/paketo-buildpacks/rfcs/issues/236

Making dependencies more customizable is a different story though, and I think that would be great to see. That connects back to https://github.com/orgs/paketo-buildpacks/discussions/58#discussioncomment-4344953

**Comment by sambhav** [comment]
Support for RHEL stacks.

**Reply by loewenstein** [reply]
For clarification: Is that a request to test with a RHEL stack or a request for Paketo to provide one?

**Reply by sambhav** [reply]
I guess if there are tests with a RHEL stack, Paketo might as well provide one. I am more interested in having RHEL support for buildpacks like ruby, python, nginx etc. But I primarily care about the above buildpacks working with a RHEL stack, not necessarily one being provided.

**Comment by till** [comment]
PHP buildpack 🫣, and smaller images across the board.

**Comment by sambhav** [comment]
Sign all buildpacks using cosign and generate an SBOM for them.

**Comment by cforce** [comment]
A buildpack with  those tools onboard to build a **fully static build  executable** (e.g including the GLIBC and not needed at runtime).
Also see  
- https://github.com/paketo-buildpacks/native-image/issues/217 
- https://github.com/spring-cloud/spring-cloud-function/issues/972
- https://github.com/oracle/graal/issues/3053

**Comment by akefirad** [comment]
A small one; [Support Structured Log Format](https://github.com/paketo-buildpacks/java/issues/874)

**Comment by jabrena** [comment]
Hi, 

I would like to know the plans for `musl` support.

Compiling a native application in a manual way, it is possible without any problem:

```xml
					<plugin>
						<groupId>org.graalvm.buildtools</groupId>
						<artifactId>native-maven-plugin</artifactId>
						<configuration>
							<buildArgs>
								<buildArg>--static</buildArg>
								<buildArg>--libc=musl</buildArg>
							</buildArgs>
						</configuration>
					</plugin>
```

But using `buildpacks`, it is not possible:

```xml
					<plugin>
						<groupId>org.springframework.boot</groupId>
						<artifactId>spring-boot-maven-plugin</artifactId>
						<configuration>
							<image>
                                                                   <buildpacks>
									<buildpack>gcr.io/paketo-buildpacks/bellsoft-liberica:9.10</buildpack>
									 <buildpack>gcr.io/paketo-buildpacks/java-native-image</buildpack>
                                                                   </buildpacks>
								<env>
									<BP_NATIVE_IMAGE>true</BP_NATIVE_IMAGE>
									<BP_JVM_VERSION>17</BP_JVM_VERSION>
									<BP_NATIVE_IMAGE_BUILD_ARGUMENTS>-J-Xmx6000m --static --libc=musl</BP_NATIVE_IMAGE_BUILD_ARGUMENTS>
									<BP_BINARY_COMPRESSION_METHOD>upx</BP_BINARY_COMPRESSION_METHOD>
									<pullPolicy>IF_NOT_PRESENT</pullPolicy>
								</env>
							</image>
						</configuration>
					</plugin>
```

Many thanks in advance

**Reply by dashaun** [reply]
Would MUSL support theoretically eliminate the glibc problems [previously identified](https://github.com/orgs/paketo-buildpacks/discussions/58#discussioncomment-4419519) by @cforce ?

**Reply by dmikusa** [reply]
Yes, these two are the same request.

**Reply by cforce** [reply]
Is there any non functional  disadvantage when compiling with MUSL (eg performance, startup time,  size of the binary, memory consumption). 
@jabrena 
As you have modified the BP_NATIVE_IMAGE_BUILD_ARGUMENTS of "spring-boot-maven-plugin" i ask myself if and how these params are merged with the ones from native-maven-plugin configuration which are there by default or configured in same pom. There are many ways to pass and override args passed to the image builder, so the question is which approach is the one .There are parameters configured by default published by spring boot (spring-boot-maven-plugin?) and  native-maven-plugin  and how is best practice to extends and not  override.

**Reply by dmikusa** [reply]
>Is there any non functional disadvantage when compiling with MUSL (eg performance, startup time, size of the binary, memory consumption).

- Static images will be larger because you're building whatever you need into the binary
- You would accept any [differences between glibc and MUSL](https://wiki.musl-libc.org/functional-differences-from-glibc.html)
- You would also need to statically build in any additional C libraries. Probably not a huge issue in this case, but if you wanted to link against another C library, it would also need to be able to be statically linked too.

**Reply by loewenstein** [reply]
I might just not get it, but isn't this more about static vs. dynamic linking and less about glibc vs. musl?

**Reply by dmikusa** [reply]
In a general sense, yes. For GraalVM, my understanding is that you have three options. Dynamic linking (what we do with base and full stacks), mostly static which only dynamically links against libc (we use this on the tiny stack) and fully static which statically links everything (we don't presently support, that's what this and the thread above are requesting). 

I don't know if it's just the way it is documented or if there are actually limitations but the [GraalVM docs](https://www.graalvm.org/22.1/reference-manual/native-image/StaticImages/) specifically list glibc and musl when explaining these options. I may occasionally mix dynamic linking with glibc and static compiling with musl for this reason, apologizes if I mess this up. The intent is to have support for a fully static binary as described in the GraalVM docs.

**Reply by jabrena** [reply]
This image was explained in Devoxx BE 2022 here:
https://www.youtube.com/watch?v=6wYrAtngIVo&ab_channel=Devoxx

![image](https://user-images.githubusercontent.com/147263/209977803-0136a4df-1961-480f-bf1c-82c06ec46de2.png)

But reviewing the official documentation, I was not able to get further details.

If you compile the image without buildpacks, it works:

```xml
					<plugin>
						<groupId>org.graalvm.buildtools</groupId>
						<artifactId>native-maven-plugin</artifactId>
						<configuration>
							<buildArgs>
								<buildArg>--static</buildArg>
								<buildArg>--libc=musl</buildArg>
							</buildArgs>
						</configuration>
					</plugin>
```

Example:
https://github.com/jabrena/101-docker/blob/7dddddc7bb6cc41c714da7c5c13058fd38214625/poliglot_docker_builds/3_java_spring_boot/build-spring-boot.sh#L13

**Reply by cforce** [reply]
So the [pom.xml](https://github.com/jabrena/101-docker/blob/main/poliglot_docker_builds/3_java_spring_boot/pom.xml#L103) <id>native</id> profile does not work with 
`mvn -Pnative spring-boot:build-image`
At least i get an issue like this

```
[INFO]     [creator]     ===> DETECTING
[INFO]     [creator]     ERROR: No buildpack groups passed detection.
[INFO]     [creator]     ERROR: Please check that you are running against the correct path.
[INFO]     [creator]     ERROR: failed to detect: no buildpacks participating
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.933 s
[INFO] Finished at: 2023-01-03T10:20:09+01:00
[INFO] ------------------------------------------------------------------------
```
Also there is an issue dependent on the native-image tag/version
https://github.com/graalvm/container/issues/66 i ran into

Did anyone created a buildpack that does the job?
[Liberica buildpack](https://github.com/paketo-buildpacks/bellsoft-liberica) would be a great candidate to participate especially for MUSL support
<BP_NATIVE_IMAGE_BUILD_ARGUMENTS>--static --libc=musl</BP_NATIVE_IMAGE_BUILD_ARGUMENTS>
https://twitter.com/fabi2tweet/status/1610295157622177796?s=20&t=kGDqivEbK_I6iLZjX0_PGw


Then i tried out with "pure" native-image  approach .. using 
@jabrena 's code
https://github.com/jabrena/101-docker/blob/7dddddc7bb6cc41c714da7c5c13058fd38214625/poliglot_docker_builds/3_java_spring_boot/build-spring-boot.sh to install (reflecting https://www.graalvm.org/latest/reference-manual/native-image/guides/build-static-executables/ guide)
and then executed the native image compile

```
TOOLCHAIN_DIR=`pwd`/x86_64-linux-musl-native
CC=${TOOLCHAIN_DIR}/bin/gcc
PATH=${TOOLCHAIN_DIR}/bin:${PATH}
javac EnvMap.java
native-image --static --libc=musl EnvMap
```
but i always stops with error messgae
```
native-image --static --libc=musl EnvMap                      
========================================================================================================================
GraalVM Native Image: Generating 'envmap' (static executable)...
========================================================================================================================
[1/7] Initializing...
                                                                                    (0.0s @ 0.16GB)
Error: Default native-compiler executable 'musl-gcc' not found via environment variable PATH
Error: To prevent native-toolchain checking provide command-line option -H:-CheckToolchain
Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception
------------------------------------------------------------------------------------------------------------------------
                         0.2s (6.0% of total time) in 6 GCs | Peak RSS: 0.48GB | CPU load: 1.87
========================================================================================================================
Failed generating 'envmap' after 3.4s.
Error: Image build request failed with exit status 1
```

musl-gcc is really not (expected) on the path but gcc via env var CC and also x86_64-linux-musl-gcc is like expected
I am using sdkman 22.3.r17-nik

With debug options...
```
 native-image --static --libc=musl -H:-CheckToolchain -H:+ReportExceptionStackTraces EnvMap    
========================================================================================================================
GraalVM Native Image: Generating 'envmap' (static executable)...
========================================================================================================================
[1/7] Initializing...
                                                                                    (0.0s @ 0.16GB)
Error: Default native-compiler executable 'musl-gcc' not found via environment variable PATH
Error: To prevent native-toolchain checking provide command-line option -H:-CheckToolchain
com.oracle.svm.core.util.UserError$UserException: Default native-compiler executable 'musl-gcc' not found via environment variable PATH
To prevent native-toolchain checking provide command-line option -H:-CheckToolchain
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:139)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker.addSkipCheckingInfo(CCompilerInvoker.java:104)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker.<init>(CCompilerInvoker.java:72)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker$LinuxCCompilerInvoker.<init>(CCompilerInvoker.java:212)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.codegen.CCompilerInvoker.create(CCompilerInvoker.java:80)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:910)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:575)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:535)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:403)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:580)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
------------------------------------------------------------------------------------------------------------------------
                         0.2s (5.9% of total time) in 6 GCs | Peak RSS: 0.47GB | CPU load: 1.91
========================================================================================================================
Failed generating 'envmap' after 3.5s.
Error: Image build request failed with exit status 
```
Then i copied `cp${TOOLCHAIN_DIR}/bin/gcc ${TOOLCHAIN_DIR}/bin/musl-gcc`

```
native-image --static --libc=musl -H:-CheckToolchain -H:+ReportExceptionStackTraces EnvMap
========================================================================================================================
GraalVM Native Image: Generating 'envmap' (static executable)...
========================================================================================================================
[1/7] Initializing...
                                                                                    (0.0s @ 0.16GB)
Error: Building images for org.graalvm.nativeimage.Platform$LINUX_AMD64 (target libc: musl) requires static JDK libraries.
Use most recent JDK from https://github.com/graalvm/labs-openjdk-*/releases
Missing libraries: java, nio, net, zip
(search path: /root/.sdkman/candidates/java/22.3.r17-nik/lib/static/linux-amd64/musl)
com.oracle.svm.core.util.UserError$UserException: Building images for org.graalvm.nativeimage.Platform$LINUX_AMD64 (target libc: musl) requires static JDK libraries.
Use most recent JDK from https://github.com/graalvm/labs-openjdk-*/releases
Missing libraries: java, nio, net, zip
(search path: /root/.sdkman/candidates/java/22.3.r17-nik/lib/static/linux-amd64/musl)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.abort(UserError.java:73)
        at org.graalvm.nativeimage.builder/com.oracle.svm.core.util.UserError.guarantee(UserError.java:97)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.NativeLibraries.initCLibraryPath(NativeLibraries.java:350)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.c.NativeLibraries.<init>(NativeLibraries.java:264)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeLibraries(NativeImageGenerator.java:1119)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.setupNativeImage(NativeImageGenerator.java:915)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:575)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:535)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:403)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:580)
        at org.graalvm.nativeimage.builder/com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:128)
------------------------------------------------------------------------------------------------------------------------
                         0.3s (6.7% of total time) in 6 GCs | Peak RSS: 0.47GB | CPU load: 1.90
========================================================================================================================
Failed generating 'envmap' after 3.6s.
Error: Image build request failed with exit status 1
````
Also see https://github.com/oracle/graal/issues/4645


I have tried arounds and  found a working solution - still figuring out what is the key (might be that it doesn't work for NIK and or you have to make  musl yourself)

```
sdk install java 22.3.r17-grl
gu install native-image
WORKDIR=`pwd`
RESULT_LIB="/staticlibs"
curl -q -L -o musl.tar.gz https://musl.libc.org/releases/musl-1.2.1.tar.gz && \
    mkdir musl && tar -xvzf musl.tar.gz -C musl --strip-components 1 && cd musl && \
    ./configure --disable-shared --prefix=${RESULT_LIB} && \
    make && make install && \
    cd / && rm -rf /muscl && rm -f /musl.tar.gz && \
    cp /usr/lib/gcc/x86_64-linux-gnu/11/libstdc++.a ${RESULT_LIB}/lib/
    #cp /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libstdc++.a ${TOOLCHAIN_DIR}/lib/
PATH="$PATH:${RESULT_LIB}/bin"
CC="musl-gcc"
cp /staticlibs/bin/musl-gcc /staticlibs/bin/x86_64-linux-musl-gcc

curl -q -L -o zlib.tar.gz https://zlib.net/zlib-1.2.13.tar.gz && \
   mkdir zlib && tar -xvzf zlib.tar.gz -C zlib --strip-components 1 && cd zlib && \
   ./configure --static --prefix=${RESULT_LIB} && \
    make && make install && \
    cd / && rm -rf /zlib && rm -f /zlib.tar.gz
cd $WORKDIR
```

**Reply by jabrena** [reply]
Yes, you can see my pipeline creating spring native apps with musl support but not using buildpacks:

- https://github.com/jabrena/101-docker/blob/main/.github/workflows/poliglot_docker_builds.yaml
- https://github.com/jabrena/101-docker/blob/main/setup-musl.sh
- https://github.com/jabrena/101-docker/blob/main/poliglot_docker_builds/3_java_spring_boot/build-spring-boot.sh

```
docker images jibber-benchmark --format '{{.Size}}\t{{.Repository}}\t{{.Tag}}\t{{.ID}}' | sed 's/ //' | sort -h -r | column -t

279MB   jibber-benchmark  buildpacks-jvm.0.0.1-SNAPSHOT     483a47cdfe58
91.3MB  jibber-benchmark  buildpacks-native.0.0.1-SNAPSHOT  9e56de567bcf
41.3MB  jibber-benchmark  buildpacks-native-static-upx.0.0.1-SNAPSHOT  9a30a86e62e7
38.7MB  jibber-benchmark  native.0.0.1-SNAPSHOT             dba7fc1dc157
18.4MB  jibber-benchmark  native-upx.0.0.1-SNAPSHOT         8801957da44e
```

To learn the way, I reviewed the examples from:
https://github.com/graalvm/graalvm-demos

**Reply by cforce** [reply]
There is an issue with nik 
sdk use java 22.3.r17-nik ->does NOT work
sdk use java 22.3.r17-grl  ->work

I expect <buildpack>gcr.io/paketo-buildpacks/bellsoft-liberica:9.10</buildpack>  would then have the same issue after we would overcome missing (detector) musl support in gcr.io/paketo-buildpacks/java-native-image.

realated mus support PR's
* https://github.com/oracle/graal/pull/3141
* https://github.com/graalvm/mx/pull/230
* https://github.com/oracle/fastr/pull/175
* https://github.com/oracle/truffleruby/pull/2223

 Where bellsoft-liberica seems define stack support for ubuntu they also seem to provide alpine with musl onbaord but which is not used as stack in buildpacks.- see https://assets.ctfassets.net/oxjq45e8ilak/3mzaJh8QxOz9VDFWeYmH01/2bdb741806000213ca775a2d4183eb28/Alpine_NIK_en_Joker.pdf

**Reply by jabrena** [reply]
From my side, I didn´t test any nik release at the moment.

**Comment by dmikusa** [comment]
Document the process for building offline images. It is possible to build offline images (where all of the buildpack dependencies are bundled into the image), however, I do not believe we currently have this process documented (100% sure for Java buildpacks, not sure for other buildpacks).  It should be a small matter to document this process and the tools.

**Comment by loewenstein** [comment]
Should we put an aligned and streamlined dependency update and release process on the roadmap? I've got mainly "using renovate for updates" in mind, but there might be more process things that would benefit from putting some effort in...

**Comment by dmikusa** [comment]
Thanks to everyone that posted feedback & upvoted issues! It is appreciated. 

We are closing this discussion to digest your feedback and will post a blog post with the summary and our 2023 roadmap shortly.
